### PR TITLE
add more docs to matrix filter aka heatmap in error analysis

### PR DESCRIPTION
**Pull request by imatiach-msft**
## Description

Add more docs to matrix filter aka heatmap in error analysis.  It looks like now all methods in erroranalysis package, both public and private, have docstrings.

## Checklist

- [ ] I have added screenshots above for all UI changes.
- [ ] I have added e2e tests for all UI changes.
- [x] Documentation was updated if it was needed.

### CODE REVIEW

**Comment by romanlutz** [comment]
Is there a way to enforce that it stays this way, i.e., failing the linter or something like that if a new method is added without docstring?

File: erroranalysis/erroranalysis/_internal/matrix_filter.py [hunk 1]
```diff
@@ -256,6 +296,22 @@ def warn_duplicate_edges(feat):
 
 
 def bin_data(df, feat, bins, quantile_binning=False):
+    """Bins the input data for the specified feature and binning method.
+
+    Uses equal width, quantile binning or specified custom bins. Custom
+    bins is used when bins is an IntervalIndex instead of a constant number.
+
+    :param df: The dataframe to bin.
```
Comment (romanlutz) [inline erroranalysis/erroranalysis/_internal/matrix_filter.py:304]:
```suggestion
    :param df: The DataFrame to bin.
```

File: erroranalysis/erroranalysis/_internal/matrix_filter.py [hunk 2]
```diff
@@ -256,6 +296,22 @@ def warn_duplicate_edges(feat):
...
+    :type df: pd.DataFrame
+    :param feat: The feature name to bin.
+    :type feat: str
+    :param bins: The number of bins or the specified bins.
+    :type bins: int or pd.IntervalIndex
+    :param quantile_binning: Whether to use quantile binning.
+    :type quantile_binning: bool
+    :returns: The binned dataframe.
```
Comment (romanlutz) [inline erroranalysis/erroranalysis/_internal/matrix_filter.py:312]:
```suggestion
    :returns: The binned DataFrame.
```

**Comment by imatiach-msft** [comment]
@romanlutz yes, when you run sphinx you can make it fail on missing docs.  I'm not sure about using a linter for docs though.

**Comment by imatiach-msft** [comment]
I think if you set warnings as errors in sphinx it will fail on missing method docs:
https://stackoverflow.com/questions/61414199/see-all-warnings-as-errors-when-building-sphinx-docs-on-readthedocs-org

File: erroranalysis/erroranalysis/_internal/surrogate_error_tree.py [hunk 3]
```diff
@@ -607,6 +607,34 @@ def create_categorical_query(method, arg, p_node_name, p_node_query,
...
+    :type feature_names: set[str]
+    :param metric: The metric to use for the current node.
+    :type metric: str
+    :param parent: The parent node.
+    :type parent: dict
+    :param side: The side of the current node from the parent, if known.
+    :type side: TreeSide
+    :param classes: The list of classes.
+    :type classes: list[str]
+    :return: The json with the node and all children added.
```
Comment (romanlutz) [inline erroranalysis/erroranalysis/_internal/surrogate_error_tree.py:635]:
```suggestion
    :return: The JSON with the node and all children added.
```

File: erroranalysis/erroranalysis/_internal/surrogate_error_tree.py [hunk 4]
```diff
@@ -607,6 +607,34 @@ def create_categorical_query(method, arg, p_node_name, p_node_query,
 def node_to_dict(df, tree, nodeid, categories, json,
                  feature_names, metric, parent=None,
                  side=TreeSide.UNKNOWN, classes=None):
+    """Converts a node and children to a dictionary that can be saved as json.
+
+    This is a method that is called on the current node and then its children
+    recursively to construct the dictionary representation.
+
+    :param df: The dataframe to use for the current node.
```
Comment (romanlutz) [inline erroranalysis/erroranalysis/_internal/surrogate_error_tree.py:615]:
```suggestion
    :param df: The DataFrame to use for the current node.
```

File: erroranalysis/erroranalysis/_internal/surrogate_error_tree.py [hunk 5]
```diff
@@ -607,6 +607,34 @@ def create_categorical_query(method, arg, p_node_name, p_node_query,
 def node_to_dict(df, tree, nodeid, categories, json,
                  feature_names, metric, parent=None,
                  side=TreeSide.UNKNOWN, classes=None):
+    """Converts a node and children to a dictionary that can be saved as json.
```
Comment (romanlutz) [inline erroranalysis/erroranalysis/_internal/surrogate_error_tree.py:610]:
```suggestion
    """Converts a node and children to a dictionary that can be saved as JSON.
```

**Comment by imatiach-msft** [comment]
I do see that we have installed:
flake8-docstrings==1.5.0
https://github.com/microsoft/responsible-ai-toolbox/blob/main/requirements-linting.txt#L8
but it doesn't fail on missing docstrings

File: erroranalysis/erroranalysis/_internal/matrix_filter.py [hunk 6]
```diff
@@ -40,13 +40,48 @@
 
 
 def compute_json_matrix(analyzer, features, filters, composite_filters):
+    """Compute a matrix of metrics for a given set of features.
+
+    :param analyzer: The error analyzer.
+    :type analyzer: BaseAnalyzer
+    :param features: A list of one or two features to compute metrics for.
```
[the remark this answers is not in the document]
Comment (imatiach-msft) [reply erroranalysis/erroranalysis/_internal/matrix_filter.py:47]: done!

File: erroranalysis/erroranalysis/_internal/matrix_filter.py [hunk 7]
```diff
@@ -40,13 +40,48 @@
...
+        saved to json.
+    :rtype: dict
+    """
     # Note: this is for backcompat for older versions
     # of raiwidgets pypi package
-    compute_matrix(analyzer, features, filters, composite_filters)
+    return compute_matrix(analyzer, features, filters, composite_filters)
 
 
 def compute_matrix(analyzer, features, filters, composite_filters,
```
[the remark this answers is not in the document]
Comment (imatiach-msft) [reply erroranalysis/erroranalysis/_internal/matrix_filter.py:62]: I think you misunderstood, num_bins is used irrespective of whether quantile binning is enabled or disabled.  If quantile_binning is false, we still bin, we just used equal width binning instead - and we always need to use the num_bin parameter.

File: erroranalysis/erroranalysis/_internal/matrix_filter.py [hunk 8]
```diff
@@ -248,6 +283,11 @@ def compute_matrix(analyzer, features, filters, composite_filters,
 
 
 def warn_duplicate_edges(feat):
+    """Alert user a feature has too many duplicate values for specified bins.
```
[the remark this answers is not in the document]
Comment (imatiach-msft) [reply erroranalysis/erroranalysis/_internal/matrix_filter.py:286]: done!

File: erroranalysis/erroranalysis/_internal/surrogate_error_tree.py (hunk 4 above)
Comment (imatiach-msft) [reply erroranalysis/erroranalysis/_internal/surrogate_error_tree.py:615]: done!

File: erroranalysis/erroranalysis/_internal/matrix_filter.py [hunk 9]
```diff
@@ -40,13 +40,48 @@
 
 
 def compute_json_matrix(analyzer, features, filters, composite_filters):
+    """Compute a matrix of metrics for a given set of features.
+
+    :param analyzer: The error analyzer.
+    :type analyzer: BaseAnalyzer
+    :param features: A list of one or two features to compute metrics for.
+    :type features: list
+    :param filters: A list of filters to apply to the data.
```
[the remark this answers is not in the document]
Comment (imatiach-msft) [reply erroranalysis/erroranalysis/_internal/matrix_filter.py:49]: done, added examples of how to call the API

File: erroranalysis/erroranalysis/_internal/matrix_filter.py [hunk 10]
```diff
@@ -40,13 +40,48 @@
...
+    :type filters: list
+    :param composite_filters: A list of composite filters to apply to the data.
```
[the remark this answers is not in the document]
Comment (imatiach-msft) [reply erroranalysis/erroranalysis/_internal/matrix_filter.py:51]: done, added examples of how to call the API

**Pull request merged**